### ISSUE TITLE
Why neg_samples can take the neighbor nodes but not only the disparate nodes

### ISSUE THREAD

**Issue by ghostHBL** [williamleif/GraphSAGE#49]
In the method of _build() in the class of SampleAndAggregate, there are codes of 
```
labels = tf.reshape(
                tf.cast(self.placeholders['batch2'], dtype=tf.int64),
                [self.batch_size, 1])
self.neg_samples, _, _ = (tf.nn.fixed_unigram_candidate_sampler(
            true_classes=labels,
            num_true=1,
            num_sampled=FLAGS.neg_sample_size,
            unique=False,
            range_max=len(self.degrees),
            distortion=0.75,
            unigrams=self.degrees.tolist()))
```
Note that we only take the output of 'sampled_candidates' when call tf.nn.fixed_unigram_candidate_sampler, the parameter of  'true_classes' seem not use, so that, why we need 'labels' in here? And the output of 'sampled_candidates' also contains the elements in the 'labels', it means that 'self.neg_samples' not only contain the elements of disparate nodes but also contains the neighbor nodes, the sample rule is only base on the degrees of each node. But in the paper of "Inductive Representation Learning on Large Graphs", what it said is "The graph-based loss function encourages nearby nodes to have similar representations while enforcing that the representations of disparate nodes are highly distinct", I think it means the 'neg_samples' in code can only contain the disparate nodes but not the neighbor nodes. Or maybe neighbor nodes should be highly distinct when they are high degrees??

**Comment by RexYing** [comment]
Hi,

Yes you are right in that the negative sampler could possibly sample nearby nodes. 
However, in the context of large graphs, as is the main focus of the paper, it is highly unlikely that the samples are neighbors of the node which we compute representations. In practice we do not find this to be a big issue. Of course if in some cases (e.g. when there are nodes with extremely large degree), you might want to modify the sampler such that neighboring nodes (or even nodes 2 hops away) are not sampled.

**Comment by ghostHBL** [comment]
In my large graph data, I also found that it will take a long long long time to random sample the neg_samples if need to make sure they are not the neighbors. So that, it's a nice choice to random sample nodes as the neg_samples without making sure they are not the neighbors since they are unlikely to be neighbors.
I will close this issue.
Thanks.
